codespan-reporting/src/lib.rs

```diff
@@ -1,4 +1,285 @@
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use codespan_reporting::Line;
+    ///
+    /// let line = Line {
+    ///     start: 2,
+    ///     number: 2,
+    ///     source: "🗻∈🌏",
+    /// };
+    ///
+    /// assert_eq!(line.column_index(0), 0);
+    /// assert_eq!(line.column_index(line.start + 0), 0);
+    /// // FIXME: Snap column index to the previous character boundary
+    /// // assert_eq!(line.column_index(line.start + 1), 0);
+    /// assert_eq!(line.column_index(line.start + 4), 1);
+    /// assert_eq!(line.column_index(line.start + 8), 3);
+    /// assert_eq!(line.column_index(line.start + 48), 3);
+    /// ```
+    pub fn column_index(&self, byte_index: usize) -> usize {
+        match byte_index.checked_sub(self.start) {
+            None => 0,
+            Some(relative_index) => {
+                let line_source = self.source.as_ref();
+
+                line_source
+                    .char_indices()
+                    .map(|(i, _)| i)
+                    .take_while(|i| *i < relative_index)
+                    .count()
+            },
+        }
+    }
+
+    /// The 1-indexed column number at the given byte index.
+    pub fn column_number(&self, byte_index: usize) -> usize {
+        self.column_index(byte_index) + 1
+    }
+}
+
+/// Files that can be used for pretty printing.
+pub trait Files {
+    type FileId: Copy + PartialEq + PartialOrd + Eq + Ord + std::hash::Hash;
+    type Origin: std::fmt::Display;
+    type LineSource: AsRef<str>;
+
+    /// The origin of a file.
+    fn origin(&self, id: Self::FileId) -> Option<Self::Origin>;
+
+    /// The line at the given index.
+    fn line(&self, id: Self::FileId, line_index: usize) -> Option<Line<Self::LineSource>>;
+
+    /// The index of the line at the given byte index.
+    fn line_index(&self, id: Self::FileId, byte_index: usize) -> Option<usize>;
+
+    /// The location of the given byte index.
+    fn location(&self, id: Self::FileId, byte_index: usize) -> Option<Location> {
+        let line_index = self.line_index(id, byte_index)?;
+        let line = self.line(id, line_index)?;
+
+        Some(Location {
+            line_number: line.number,
+            column_number: line.column_number(byte_index),
+        })
+    }
+}
+
+/// A single source file.
+///
+/// This is useful for simple language tests, but it might be worth creating a
+/// custom implementation when a language scales beyond a certain size.
+#[derive(Debug, Clone)]
+pub struct SimpleFile<Origin, Source> {
+    /// The origin of the file.
+    origin: Origin,
+    /// The source code of the file.
+    source: Source,
+    /// The starting byte indices in the source code.
+    line_starts: Vec<usize>,
+}
+
+fn line_starts<'a>(source: &'a str) -> impl 'a + Iterator<Item = usize> {
+    std::iter::once(0).chain(source.match_indices('\n').map(|(i, _)| i + 1))
+}
+
+impl<Origin, Source> SimpleFile<Origin, Source>
+where
+    Origin: std::fmt::Display,
```

Comment:
I like this API. Currently I store both the absolute path in the `Source` struct and the filename in Files, this would allow me to store both in Origin instead which semantically makes more sense :)